### PR TITLE
feat(finetune): support multiple datasets

### LINKED ISSUE #825: [REQ] support multiple fine-tuning datasets
Parent: #214

Depends on #820, #824, and #823.

## Summary

Allow more than one SFT dataset and concatenate normalized records deterministically. Dataset order in the aikitfile and row order within each source must be preserved before trainer-level shuffling.

```yaml
datasets:
  - source: acme/support-conversations
    type: messages
  - source: acme/legacy-conversations
    type: sharegpt
```

## Motivation

Current layers disagree about multiple datasets:

- Go stores `[]Dataset`.
- Build validation rejects more than one.
- LLB serialization includes the whole slice.
- Python reads only `datasets[0]`.

Removing the Go validation alone would silently ignore every dataset after the first. Safe composition requires loading and normalizing each source before combining compatible schemas and loss semantics.

## Proposed changes

- Remove the one-dataset validation only after Python iterates over all entries.
- Load and validate every dataset in YAML order.
- Include dataset index/type in errors without exposing sensitive source details.
- Normalize each dataset independently to a canonical trainer schema.
- Initially define compatible groups:

```text
Full-sequence text:
  alpaca, text, messages(all), sharegpt(all)

Completion-only:
  prompt-completion

Response-only chat:
  messages(response), sharegpt(response)
```

- Require all datasets in a job to use the same canonical schema and loss behavior.
- Reject incompatible groups rather than coercing them.
- Remove source-specific columns after normalization.
- Concatenate deterministically in aikitfile order, preserving row order within each dataset.
- Reject empty datasets instead of skipping them.
- Preserve current single-dataset behavior exactly.
- Keep all dataset identities/options in the training config and cache key.

## Acceptance criteria

- Every configured dataset is loaded exactly once; Python no longer reads only `datasets[0]`.
- Records appear in aikitfile order before trainer shuffling.
- Compatible full-sequence, prompt-completion, and response-only groups can be concatenated.
- Incompatible schemas/loss modes fail with clear dataset indexes and types.
- Empty datasets fail.
- Existing one-dataset Alpaca and other supported configurations behave identically.
- Changing only the second or later dataset changes the training LLB digest.
- A one-step GPU run using two small compatible datasets exports a valid GGUF.

## Tests

- Go tests accepting multiple supported datasets.
- Python tests proving every source is loaded and later entries are not ignored.
- Deterministic ordering tests using real `datasets.Dataset` objects.
- Compatibility tests for Alpaca+text, messages+ShareGPT, multiple prompt-completion datasets, and response-only messages+ShareGPT.
- Rejection tests for mixed canonical schemas/loss modes and empty datasets.
- LLB digest test changing only a later dataset.
- GPU smoke test with two sources.

## Out of scope

- Weighted sampling, random interleaving, or per-dataset weights.
- Oversampling/undersampling or deduplication.
- Per-dataset loss settings.
- Automatic coercion between incompatible schemas.
- Streaming or evaluation datasets.
- Multiple DPO/ORPO/KTO preference datasets.

### PR DESCRIPTION
**Pull request by sozercan**
Stacked on #833. This PR targets `issue-824` and should be merged after #833.

## Summary

- support multiple SFT datasets with deterministic YAML-order loading and source-row-order concatenation
- validate every source independently and reject empty or incompatible datasets before model allocation
- enforce compatible full-sequence, prompt-completion, and response-only semantic groups
- preserve each source's loader, split, subset, revision, checksum, and URL-redaction behavior
- normalize sources to canonical string features before ordered `datasets.concatenate_datasets`
- preserve the existing single-source path without concatenation
- aggregate prepared-data fingerprints across sources while retaining duplicates and tolerating BFD reorder
- apply response-only masking once to the combined trainer
- invalidate training/export when later datasets change or reorder without invalidating dependency installation
- add exact Datasets 4.3.0 unit coverage, two-source GPU smoke coverage, fixtures, and documentation

## Tests

- `make test`
- Python suite with exact `datasets==4.3.0` and `pyarrow==25.0.0` (141 tests, no skips)
- focused Go config/build/LLB tests
- exact-lock check with uv 0.12.1
- changed-lines `golangci-lint` (0 issues)
- actionlint/ShellCheck integration and workflow fixture rendering
- `go mod tidy` clean
- independent compatibility/order/cache-boundary review

The `multiple-datasets-smoke` GPU fine-tune workflow will be dispatched on this branch.

Closes #825

## GPU validation

- `multiple-datasets-smoke` passed the two-source fine-tune, GGUF export, image build, and inference path on final branch head `afede84`: [run 31177395921](https://github.com/kaito-project/aikit/actions/runs/31177395921).
- The GitHub-hosted unit-test job also passes with the exact Datasets/PyArrow test environment after switching dependency installation to a repository-local uv virtual environment.

### CODE REVIEW

File: pkg/finetune/target_unsloth.py [hunk 1]
```diff
@@ -3758,12 +4122,94 @@ def train_model(
...
+    if response_only or validate_all_full_sequence_records:
+        if len(canonical_datasets) == 1 and has_chat_dataset:
+            messages_token_fingerprints = canonical_chat_fingerprints
+        else:
+            first_record = next(iter(canonical_datasets[0]))
+            full_sequence_add_special_tokens = (
+                messages_unsloth_add_special_tokens(
+                    tokenizer,
+                    first_record["text"],
+                )
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/finetune/target_unsloth.py:4803]: Good catch. The pinned Unsloth 2026.8.3 SFT path derives one `add_special_tokens` value from the first canonical record. We reproduced both orderings: Alpaca-first duplicated BOS on the chat record, while chat-first omitted tokenizer-added BOS from the Alpaca record.

Commit `8e2ca8e` fixes this by comparing each full-sequence source's effective token IDs under its source-specific policy with the combined dataset-wide policy before concatenation. It rejects only differences that change token IDs, so harmless policy differences remain supported. Regression tests cover both source orderings and the no-op case, and the docs now describe the tokenizer-boundary requirement.

File: pkg/finetune/target_unsloth.py [hunk 2]
```diff
@@ -3633,97 +3974,157 @@ def train_model(
...
-            max_seq_length=max_seq_length,
-            source=dataset_spec.source,
-        )
+
+    prompt_add_special_tokens = None
+    if completion_only:
+        first_prompt_record = next(iter(source_datasets[0]))
+        prompt_add_special_tokens = prompt_completion_add_special_tokens(
+            tokenizer,
+            first_prompt_record["prompt"],
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/finetune/target_unsloth.py]: Good catch. This is valid and reproduces in both source orderings: a plain-prompt-first job can duplicate BOS on a later explicit-BOS prompt, while the reverse ordering can omit tokenizer-added BOS from a later plain prompt. The previous expected fingerprint was derived with the same combined policy, so it accepted the altered boundary.

Commit `c57f88c` now compares each prompt-completion source's effective retained token IDs and completion mask under its source-specific `add_special_tokens` policy with the combined dataset policy selected from the first source. It rejects only effective token or mask changes, so no-op policy differences remain supported. Regression tests cover both orderings, and the behavior is documented.
